File: partiql-ast/src/main/java/org/partiql/ast/Exclude.java
package org.partiql.ast;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

/**
 * Represents PartiQL's EXCLUDE clause.
 */
@Builder(builderClassName = "Builder")
@EqualsAndHashCode(callSuper = false)
public final class Exclude extends AstNode {
    @NotNull
    private final List<ExcludePath> excludePaths;

    public Exclude(@NotNull List<ExcludePath> excludePaths) {
        this.excludePaths = excludePaths;
    }

    @NotNull
    @Override
    public List<AstNode> getChildren() {
        return new ArrayList<>(excludePaths);
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitExclude(this, ctx);
    }

    @NotNull
    public List<ExcludePath> getExcludePaths() {
        return this.excludePaths;
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/Exclude.kt
package org.partiql.ast.v1

/**
 * TODO docs, equals, hashcode
 */
public class Exclude(
    @JvmField
    public var excludePaths: List<ExcludePath>,
) : AstNode() {
    public override fun children(): Collection<AstNode> {
        val kids = mutableListOf<AstNode?>()
        kids.addAll(excludePaths)
        return kids.filterNotNull()
    }

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
        visitor.visitExclude(this, ctx)
}
